News 212: add review club summary (#24)
harding/bitcoinops.github.io: _posts/en/newsletters/2022-08-10-newsletter.md

meeting, highlighting some of the important questions and answers.  Click on a
question below to see a summary of the answer from the meeting.*

> [Decouple validation cache initialization from ArgsManager][review club 25527]
> is a PR by Carl Dong that is part of the [libbitcoinkernel project][].
> The goal of this project is to decouple Bitcoin Core's consensus engine from
> non-consensus code.

Reviewer: Instead of summarizing libbitcoinkernel, summarize n25527. For example,
```suggestion
[Decouple validation cache initialization from ArgsManager][review club 25527]
is a PR by Carl Dong that separates node configuration logic from the
initialization of signature and script caches. It is part of the [libbitcoinkernel project][].
```

When a new block arrives, the most computationally expensive part of validating
it is validating its transactions; this is part of the consensus engine.
Since most of a block's transactions have already
been seen and validated (when added to the mempool), block validation can
be sped up by caching the (successful) script and signature verifications
that were done earlier. These caches are also part of the
consensus engine. The size of caches can be specified using a 
configuration option, `-maxsigcachesize`. This setting, like all
configuration settings, is stored within Bitcoin Core's global

> [`ArgsManager`][] data structure.

Reviewer: I think the idea here is to summarize the meeting itself rather than the review club notes, so this background isn't necessary. If the information was discussed in the meeting, maybe you can incorporate some of this into the question/answer portion?

My approach to writing these summaries was to split the log into sections (usually by question, but often by "topic") since it's not always 100% focused on the questions, pick the 4-6 sections with the most discussion, and summarize them.

Reviewer: For example, you can fit most of this pretty neatly into the "what are the validation caches? Why would they belong in src/kernel vs src/node?" question. https://bitcoincore.reviews/25527#l-45

{% include functions/details-list.md
  q0="Why or why not should `ArgsManager` belong in `src/kernel` versus `src/node`?"
  a0="It's undesirable for the consensus engine to access `ArgsManager` for
any reason, because it could, in principle, allow per-node configuration options
to modify the behavior of the consensus engine. This could lead to
a loss of consensus between nodes."

>   a0link="https://bitcoincore.reviews/25527#l-38"

Reviewer 2: ```suggestion
  a0link="https://bitcoincore.reviews/25527#l-35"
```

  q1="What tools do you use for “code archeology” to understand the background of why
a value exists?"
  a1="`git blame filename`, `git log -p filename`, or enter
`commit:<commit-hash>` into the
[pulls](https://github.com/bitcoin/bitcoin/pulls) page."

>   a1link="https://bitcoincore.reviews/25527#l-140"

Reviewer 2: ```suggestion
  a1link="https://bitcoincore.reviews/25527#l-132"
```

  q2="This PR changes the type of `signature_cache_bytes` and
`script_execution_cache_bytes` from `int64_t` to `size_t`.
What is the difference between `int64_t`, `uint64_t`, and `size_t`,
and why should a `size_t` hold these values?"
  a2="The `int64_t` and `uint64_t` types are 64-bits (signed and unsigned,
respectively) across all platforms and compilers. The `size_t` type
is an unsigned integer that's
guaranteed to be able to hold the length (in bytes) of any object
in memory, including an array.
The `size_t` type can be either 32 bits on a 32-bit memory system or
64 bits on a 64-bit memory system. A `size_t` can also hold an array index,
which is one of its most common uses.
It also matches the size of a memory pointer."

>   a2link="https://bitcoincore.reviews/25527#l-171"

Reviewer 2: ```suggestion
  a2link="https://bitcoincore.reviews/25527#l-163"
```
